### PR TITLE
first pass at a migration script for TDS

### PR DESCRIPTION
**Pull request by dgauldie**
# Description

* Include a summary of the changes and the related issue. 
* Include relevant motivation and context.

Resolves #(issue)

### CODE REVIEW

**Comment by mwdchang** [comment]
No indices in this pass?

File: packages/server/build.gradle [hunk 1]
```diff
@@ -27,6 +27,7 @@ dependencyManagement {
 
 
 apply plugin: 'cz.habarta.typescript-generator'
+apply plugin: 'idea'
```
[the remark this answers is not in the document]
Comment (dgauldie) [reply packages/server/build.gradle:30]: nothing

[the remark this answers is not in the document]
Comment (dgauldie) [reply packages/server/build.gradle:30]: All it does is it generates a module file and an intellij project and workspace file in the .idea folder. It also adds Java configurations. all of it is in the .idea folder and it is git ignored.

[the remark this answers is not in the document]
Comment (dgauldie) [reply packages/server/build.gradle:30]: helps with intellij

File: packages/tds-migration/settings.gradle [hunk 2]
```diff
@@ -0,0 +1 @@
+rootProject.name = 'terarium'
```
[the remark this answers is not in the document]
Comment (dgauldie) [reply packages/tds-migration/settings.gradle:1]: nope

File: packages/tds-migration/src/main/java/software/uncharted/terarium/tds/migration/callbacks/FlywayMigrationCompleteCallback.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  @Override
+  public boolean canHandleInTransaction(final Event event, final Context context) {
+    return true;
+  }
+
+  @Override
+  public void handle(final Event event, final Context context) {
+    if (event == Event.AFTER_MIGRATE) {
```
Comment (YohannParis) [inline packages/tds-migration/src/main/java/software/uncharted/terarium/tds/migration/callbacks/FlywayMigrationCompleteCallback.java:27]: Why not use `supports(event)`.

File: packages/tds-migration/src/main/resources/application.properties [hunk 4]
```diff
@@ -0,0 +1,15 @@
+########################################################################################################################
+# Database configuration
+########################################################################################################################
+spring.datasource.url=jdbc:postgresql://localhost:5430/tds
```
Comment (YohannParis) [inline packages/tds-migration/src/main/resources/application.properties:4]: Does `localhost` will work on k8 on AWS?

File: packages/tds-migration/src/main/resources/db/data/R__project.sql [hunk 5]
```diff
@@ -0,0 +1,10 @@
+INSERT INTO project
+    (id, name, description, created_on, updated_on, deleted_on)
+VALUES
+    (1, 'ASKEM Epi Demo Project', 'ASKEM Epi Demo Project', '2023-01-01 00:00:00', '2023-01-01 00:00:00', NULL)
+ON CONFLICT DO NOTHING;
+INSERT INTO project
+    (id, name, description, created_on, updated_on, deleted_on)
+VALUES
+    (2, 'ASKEM Climate Demo Project', 'ASKEM Climate Demo Project', '2023-01-01 00:00:00', '2023-01-01 00:00:00', NULL)
+ON CONFLICT DO NOTHING;
```
Comment (YohannParis) [inline packages/tds-migration/src/main/resources/db/data/R__project.sql:10]: You can bundle into one `INSERT` call.
```suggestion
INSERT INTO project
    (id, name, description, created_on, updated_on, deleted_on)
VALUES
    (1, 'ASKEM Epi Demo', 'ASKEM Epi Demo', '2023-01-01 00:00:00', '2023-01-01 00:00:00', NULL),
    (2, 'ASKEM Climate Demo', 'ASKEM Climate Demo', '2023-01-01 00:00:00', '2023-01-01 00:00:00', NULL)
ON CONFLICT DO NOTHING;
```

File: settings.gradle [hunk 6]
```diff
@@ -8,5 +8,5 @@ pluginManagement {
 
 rootProject.name = "terarium"
 
-include "packages:server"
-
+include ":packages:server"
+include ":packages:tds-migration"
```
Comment (YohannParis) [inline settings.gradle:12]:
```suggestion
include ":packages:db-migration"
```

File: packages/tds-migration/src/main/java/software/uncharted/terarium/tds/migration/callbacks/FlywayMigrationCompleteCallback.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.context.ApplicationContext;
+
+@Slf4j
+public record FlywayMigrationCompleteCallback(
+  ApplicationContext applicationContext) implements Callback {
+
+  @Override
+  public boolean supports(final Event event, final Context context) {
+    return event == Event.AFTER_MIGRATE;
+  }
```
[the remark this answers is not in the document]
Comment (mwdchang) [reply packages/tds-migration/src/main/java/software/uncharted/terarium/tds/migration/callbacks/FlywayMigrationCompleteCallback.java:18]: I think this is kind of a case where you don't have dynamic-dispatch in Java 😄

File: packages/tds-migration/src/main/resources/application.properties (hunk 4 above)
Comment (dgauldie) [reply packages/tds-migration/src/main/resources/application.properties:4]: running this on kubernetes you would overwrite just like we do in hmi-server

File: packages/tds-migration/src/main/resources/db/data/R__project.sql (hunk 5 above)
Comment (dgauldie) [reply packages/tds-migration/src/main/resources/db/data/R__project.sql:10]: that's all or nothing. The original way, we can have 1 conflict and not the other.

File: packages/tds-migration/src/main/java/software/uncharted/terarium/tds/migration/callbacks/FlywayMigrationCompleteCallback.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (YohannParis) [reply packages/tds-migration/src/main/java/software/uncharted/terarium/tds/migration/callbacks/FlywayMigrationCompleteCallback.java]: Okay, if Java is fine with an unused argument.

File: packages/tds-migration/src/main/resources/db/data/R__project.sql (hunk 5 above)
Comment (YohannParis) [reply packages/tds-migration/src/main/resources/db/data/R__project.sql]: That makes sense, but to be honest, this is the only script that will ever add those projects, so either both exist or neither.
It doesn't matter, this is me being pedantic for no good reasons.

**Review by YohannParis — APPROVED** [review]
You have no idea how the renaming from `tds` to `db` makes me happy 😄

File: packages/tds-migration/README.md [hunk 8]
```diff
@@ -0,0 +1,15 @@
+# TDS Database Migration
+
+Once run, this application will update the TDS database to the latest version (if needed) and then exit.
```
[the remark this answers is not in the document]
Comment (dgauldie) [reply packages/tds-migration/README.md]: fixed

File: packages/tds-migration/src/main/java/software/uncharted/terarium/tds/migration/callbacks/FlywayMigrationCompleteCallback.java (hunk 3 above)
Comment (dgauldie) [reply packages/tds-migration/src/main/java/software/uncharted/terarium/tds/migration/callbacks/FlywayMigrationCompleteCallback.java]: done

**Pull request merged**